File: JAVA/Graph_prims.java

import java.util.*; 
class MST { 
	private static final int V = 5; 
	int minKey(int key[], Boolean mstSet[]) 
	{ 
		int min = Integer.MAX_VALUE, min_index = -1; 
		for (int v = 0; v < V; v++) 
			if (mstSet[v] == false && key[v] < min) { 
				min = key[v]; 
				min_index = v; 
			} 
		return min_index; 
	} 
	void printMST(int parent[], int graph[][]) 
	{ 	int minCost = 0;
		System.out.println("Edge \tWeight"); 
		for (int i = 1; i < V; i++)
			{ 
			System.out.println(parent[i] + " - " + i + "\t" + graph[i][parent[i]]);
			minCost += graph[i][parent[i]];
			}
		System.out.println("Minimum cost of spanning tree by Prim's algorithm is "+ minCost);
	} 
	void primMST(int graph[][]) 
	{ 		 
		int parent[] = new int[V]; 	 
		int key[] = new int[V]; 
		Boolean mstSet[] = new Boolean[V]; 
		for (int i = 0; i < V; i++) { 
			key[i] = Integer.MAX_VALUE; 
			mstSet[i] = false; 
		} 
		key[0] = 0; 
		parent[0] = -1;  
		for (int count = 0; count < V - 1; count++) { 
			int u = minKey(key, mstSet); 
			mstSet[u] = true; 
			for (int v = 0; v < V; v++) 
				if (graph[u][v] != 0 && mstSet[v] == false && graph[u][v] < key[v]) { 
					parent[v] = u; 
					key[v] = graph[u][v]; 
				} 
		} 
		printMST(parent, graph); 
	} 

	public static void main(String[] args) 
	{ 
		MST t = new MST(); 
		int n;
		Scanner sc = new Scanner(System.in);
		System.out.println("Enter the number of Vertices of graph"); 
		n = sc.nextInt();
		int graph[][] = new int[50][50];
		int i = 0;
		int j= 0;
		System.out.println("Enter the Adjacency Matrix of order"+n+"x"+n+"");
		for (i=0;i<n;i++) {
			for (j = 0; j<n; j++){
				graph[i][j] = sc.nextInt();
		}}
		long start = System.nanoTime();
		t.primMST(graph);
		long end = System.nanoTime(); 
		double execution = end - start;
    	System.out.println("\nExecution time: " + execution/1000000 + " milliseconds");
	} 
} 

File: Python/graph_prims.py
import sys
import time

class Graph:
    def __init__(self,vertices):
        self.V = vertices
        self.graph = [[0 for row in range(vertices)]for column in range(vertices)]

    def printMST(self, parent):
        minCost = 0
        print("\nThe edges in minimum spanning tree of given graph are\n")
        print("Edge\tWeight")
        for i in range(1, self.V):
            print(parent[i], "--", i,"\t" , self.graph[i][parent[i]])
            minCost += self.graph[i][parent[i]]

        print(f"Minimum cost of spanning tree by Prim's algorithm is =", minCost)

    def minKey(self, key, mstSet):

        min = sys.maxsize

        for v in range(self.V):
            if key[v]< min and mstSet[v] == False:
                min = key[v]
                min_index = v

        return min_index

    def primMST(self):

        key = [sys.maxsize] * self.V
        parent = [None]* self.V
        mstSet = [False]* self.V
        key[0] = 0
        parent[0] = -1

        for x in range(self.V):
            u = self.minKey(key, mstSet)

            mstSet[u] = True

            for v in range(self.V):
                if self.graph[u][v] > 0 and key[v]> self.graph[u][v] and mstSet[v] == False:
                    key[v] = self.graph[u][v]
                    parent[v] = u

        self.printMST(parent)

n = int(input("Enter number of vertices of the graph "))
g = Graph(n)
print("Enter the adjacency matrix of the Graph \n")
for i in range(n):
    print("Enter", n ,"values in row",i+1)
    for j in range(n):
        g.graph[i][j]=int(input())

start = time.time()
g.primMST()
end = time.time()
print(end = "\n")
print(f"Runtime of Prims's algorithm is {(end-start)*1000} milliseconds")












        
